File: sorting/perl/shell.pl
#!/usr/bin/perl

sub shell  
{ 
    my @tosort = @_;
    my $inc = int(($#tosort+1) / 2);
    while ($inc > 0) {
        foreach my $i (0 .. $#tosort) {
            my $temp = $tosort[$i];
            while ($i >= $inc && $tosort[$i-$inc] > $temp) {
                $tosort[$i] = $iary[$i-$inc];
                $i -= $inc;
            }
            $tosort[$i] = $temp;
        }    
        if ($inc == 2) {
            $inc = 1;
        } else { 
           $inc *= (5.0 / 11);
           $inc = int($inc);
        }
    }
    return @tosort;
}

File: sorting/ruby/shell.rb
#!/usr/bin/bash

class Array
  def shellsort!
    inc = length / 2
    while inc != 0
      each_with_index do |el, i|
        while i >= inc and self[i - inc] > el
          self[i] = self[i - inc]
          i -= inc
        end
        self[i] = el
      end
      inc = (inc == 2 ? 1 : (inc * 5.0 / 11).to_i)
    end
  end
end
